Backend/lib/yudhishthira/src/Lib/Yudhishthira/Event/KaalChakra.hs

```diff
@@ -0,0 +1,156 @@
+import qualified Data.Aeson.QQ.Simple as AQ
+import qualified Data.Aeson.Types as A
+import Data.List (nub)
+import qualified Data.List as DL
+import qualified Data.Map as M
+import Data.OpenApi (ToSchema)
+import Data.Scientific (toRealFloat)
+import qualified Data.Text as T
+import qualified JsonLogic
+import Kernel.Prelude
+import qualified Kernel.Storage.ClickhouseV2 as CH
+import qualified Kernel.Types.APISuccess
+import Kernel.Types.Id
+import Kernel.Utils.Common
+import Lib.Yudhishthira.Storage.Beam.BeamFlow
+import Lib.Yudhishthira.Storage.Queries.ChakraQueries as QChakraQueries
+import qualified Lib.Yudhishthira.Storage.Queries.ChakraQueries as SQCQ
+import qualified Lib.Yudhishthira.Storage.Queries.NammaTag as QNT
+import qualified Lib.Yudhishthira.Storage.Queries.NammaTag as QNammaTag
+import qualified Lib.Yudhishthira.Storage.Queries.UserData as QUserData
+import qualified Lib.Yudhishthira.Storage.Queries.UserTag as QUserTag
+import Lib.Yudhishthira.Tools.Error
+import Lib.Yudhishthira.Tools.Utils
+import qualified Lib.Yudhishthira.Types
+import qualified Lib.Yudhishthira.Types.ChakraQueries
+import qualified Lib.Yudhishthira.Types.NammaTag as DNT
+import qualified Lib.Yudhishthira.Types.UserData as DUserData
+import qualified Lib.Yudhishthira.Types.UserTag as DUserTag
+import Servant hiding (throwError)
+
+kaalChakraEvent ::
+  (BeamFlow m r, CH.HasClickhouseEnv CH.APP_SERVICE_CLICKHOUSE m) =>
+  Lib.Yudhishthira.Types.Chakra ->
+  m ()
+kaalChakraEvent chakra = do
+  chakraQueries <- QChakraQueries.findAllByChakra chakra
+  chakraQueriesResults <- forM chakraQueries $ \chakraQuery -> do
+    queryResult <- runQueryRequest chakraQuery
+    pure (chakraQuery.queryName, queryResult)
+  userDataList <- buildUserDataList chakra chakraQueriesResults
+  QUserData.createMany userDataList
+
+  tags <- QNammaTag.findAllByChakra chakra
+  forM_ tags $ \tag -> do
+    case tag.rule of
+      Lib.Yudhishthira.Types.RuleEngine logic -> do
+        forM_ userDataList $ \userData -> do
+          let eTagValueObj = JsonLogic.jsonLogicEither logic userData.userDataValue
+          case eTagValueObj of
+            Left err -> logError $ "Could not apply rule to data: userId: " <> show userData.userId <> "; tag: " <> show tag.name <> "; data: " <> show userData.userDataValue <> "; error: " <> show err
+            Right tagValueObj -> do
+              case parseTagValue tagValueObj tag.possibleValues of
+                Nothing -> logError ("Value is not allowed: " <> show tagValueObj <> "; tag: " <> tag.name)
+                Just tagValue -> upsertUserTag userData.userId tag.name tagValue
+      Lib.Yudhishthira.Types.LLM _ -> throwError (InternalError "LLM is not implemented")
+
+upsertUserTag ::
+  (BeamFlow m r) =>
+  Id Lib.Yudhishthira.Types.User ->
+  Text ->
+  Lib.Yudhishthira.Types.TagValue ->
+  m ()
+upsertUserTag userId tagName tagValue = do
+  mbUserTag <- QUserTag.findByPrimaryKey tagName userId
```

Comment:
We should update tag in `person` table of of driver, instead of creating new table.